avutil/hwcontext_qsv: fix D3D11VA hwmap errors

tong1wu opened the pull request:
The mfxHDLPair information should be put into texture_infos when
deriving from qsv context. Now fix it to make the qsv->d3d11va zero copy
work properly.

xhaihao commented on libavutil/hwcontext_d3d11va.c:
```diff
@@ -344,6 +344,22 @@ static int d3d11va_create_staging_texture(AVHWFramesContext *ctx)
+        }
+    }
+
     D3D11_TEXTURE2D_DESC texDesc = {
```
Please do not mix declarations and statements

xhaihao commented on libavutil/hwcontext_qsv.c:
```diff
@@ -806,9 +806,15 @@ static int qsv_frames_derive_from(AVHWFramesContext *dst_ctx,
+                mfxHDLPair* pair = (mfxHDLPair*)src_hwctx->surfaces[i].Data.MemId;
+                dst_hwctx->texture_infos[i].texture = (ID3D11Texture2D*)pair->first;
+                dst_hwctx->texture_infos[i].index = (intptr_t)pair->second;
+            }
```
Could you add a sample command in the git log to demonstrate the issue ?

tong1wu replied:
ok, I will update the patchset.

tong1wu replied:
> Could you add a sample command in the git log to demonstrate the issue ?

I have resubmitted them, thx.

wenbinc-Bin commented on libavutil/hwcontext_qsv.c:
```diff
@@ -1287,6 +1311,7 @@ static int qsv_frames_derive_to(AVHWFramesContext *dst_ctx,
         return AVERROR(ENOSYS);
     }
 
+    s->child_frames_ref = av_buffer_ref(dst_ctx->internal->source_frames);
```
You need a failure check here.

wenbinc-Bin commented on libavutil/hwcontext_qsv.c:
```diff
@@ -900,9 +910,16 @@ static int qsv_map_from(AVHWFramesContext *ctx,
 
        if (child_frames_ctx->device_ctx->type == AV_HWDEVICE_TYPE_D3D11VA) {
+#if CONFIG_D3D11VA
+           AVD3D11VAFramesContext* child_frames_hwctx = child_frames_ctx->hwctx;
```
This indent is not aligned.

wenbinc-Bin commented on libavutil/hwcontext_qsv.c:
```diff
@@ -806,9 +806,19 @@ static int qsv_frames_derive_from(AVHWFramesContext *dst_ctx,
-            dst_hwctx->texture = (ID3D11Texture2D*)pair->first;
+            dst_hwctx->texture_infos = av_calloc(src_hwctx->nb_surfaces,
+                                            sizeof(*dst_hwctx->texture_infos));
+            for (i = 0; i < src_hwctx->nb_surfaces; i++) {
```
I think we should move this loop after this line
`dst_hwctx->BindFlags = qsv_get_d3d11va_bind_flags(src_hwctx->frame_type);`

tong1wu replied:
ok I will update

wenbinc-Bin commented:
LGTM Thanks.

Commit:
avutil/hwcontext_d3d11va: add a format check for staging texture

The texDesc.Format needs to be filled in with a corresponding format. I
add a format check to initialize the format in case sometimes the
ctx->internal->priv is not initialized, such as during the hwmap
process.

$ ffmpeg.exe -y -hwaccel qsv -init_hw_device d3d11va=d3d11 \
-init_hw_device qsv=qsv@d3d11 -c:v h264_qsv \
-i input.h264 -vf "hwmap=derive_device=d3d11va,format=d3d11,hwdownload,format=nv12" \
-f null -

Signed-off-by: Tong Wu <tong1.wu@intel.com>

Commit:
avutil/hwcontext_d3d11va: fix the uninitialized texture bindflag

When uploading rawvideos using d3d11va hardware framecontext, the bindflag
is not initialized and will cause creating texture failure. Now fix it,
assign it the value of D3D11_BIND_RENDER_TARGET.

Fixes:
$ ffmpeg.exe -init_hw_device d3d11va=d3d11 -f lavfi -i yuvtestsrc -vf \
"format=nv12,hwupload=extra_hw_frames=16" -f null -

Signed-off-by: Tong Wu <tong1.wu@intel.com>

Commit:
avutil/hwcontext_qsv: derive QSV frames to D3D11VA frames

Fixes:
$ ffmpeg.exe -y -hwaccel qsv -init_hw_device d3d11va=d3d11 \
-init_hw_device qsv=qsv@d3d11 -c:v h264_qsv -i input.h264 \
-vf "hwmap=derive_device=d3d11va,format=d3d11" -f null -

Signed-off-by: Tong Wu <tong1.wu@intel.com>

tong1wu committed:
avutil/hwcontext_qsv: map QSV frames to D3D11VA frames

When input is a rawvideo, after mapping QSV frames to D3D11VA frames,
the output will have green frames. Now fix it.

Fixes:
$ ffmpeg.exe -y -hwaccel qsv -init_hw_device d3d11va=d3d11 \
-init_hw_device qsv=qsv@d3d11 -s:v WxH -pix_fmt nv12 -i input.yuv \
-vf "format=nv12,hwupload=extra_hw_frames=16,\
hwmap=derive_device=d3d11va,format=d3d11,hwdownload,format=nv12" \
-f rawvideo output.yuv

Signed-off-by: Tong Wu <tong1.wu@intel.com>

tong1wu committed:
avutil/hwcontext_qsv: map D3D11VA frames to QSV frames

Fixes:
$ ffmpeg.exe -y -hwaccel d3d11va -init_hw_device d3d11va=d3d11 \
-s:v WxH -pix_fmt nv12 -i input.yuv -vf "format=nv12,\
hwupload=extra_hw_frames=16,hwmap=derive_device=qsv,format=qsv,\
hwdownload,format=nv12" -f null -

Signed-off-by: Tong Wu <tong1.wu@intel.com>

tong1wu commented:
@xhaihao I have split them into 5 commits. The first 3 commits are maintained as discussed and I added the last 2 commits. Please help to review the last 2 thanks.

tong1wu commented:
closed. Merged into upstream.
